pkg/lib/infra/redis/hub.go

```diff
@@ -0,0 +1,244 @@
+package redis
+
+import (
+	"context"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-redis/redis/v8"
+
+	"github.com/authgear/authgear-server/pkg/lib/config"
+	"github.com/authgear/authgear-server/pkg/util/log"
+)
+
+const (
+	RedisReadTimeout = 1 * time.Second
+)
+
+type SubscriberID struct {
+	Counter int64
+}
+
+func (i *SubscriberID) Get() int64 {
+	for {
+		curr := atomic.LoadInt64(&i.Counter)
+		if atomic.CompareAndSwapInt64(&i.Counter, curr, curr+1) {
+			return curr
+		}
+	}
+}
+
+type SubscriberKey struct {
+	ConnKey     string
+	ChannelName string
+}
+
+type Subscriber struct {
+	SubscriberID int64
+	Chan         chan *redis.Message
+}
+
+// Hub aims to multiplex multiple subscription to Redis PubSub channel over a single connection.
+type Hub struct {
+	Pool         *Pool
+	SubscriberID *SubscriberID
+	Logger       *log.Logger
+
+	mutex      sync.RWMutex
+	pubsub     map[string]*redis.PubSub
+	subscriber map[SubscriberKey][]Subscriber
+}
+
+func NewHub(pool *Pool, lf *log.Factory) *Hub {
+	h := &Hub{
+		Pool:         pool,
+		SubscriberID: &SubscriberID{},
+		Logger:       lf.New("redis-hub"),
+
+		subscriber: make(map[SubscriberKey][]Subscriber),
+		pubsub:     make(map[string]*redis.PubSub),
+	}
+	return h
+}
+
+func (h *Hub) Subscribe(
+	cfg *config.RedisConfig,
+	credentials *config.RedisCredentials,
+	channelName string,
+) (chan *redis.Message, func(), error) {
+	connKey := credentials.ConnKey()
+
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
+	pubsub := h.getPubSub(cfg, credentials)
+	id := h.SubscriberID.Get()
+
+	key := SubscriberKey{
+		ConnKey:     connKey,
+		ChannelName: channelName,
+	}
+
+	c := make(chan *redis.Message)
+
+	h.subscriber[key] = append(h.subscriber[key], Subscriber{
+		SubscriberID: id,
+		Chan:         c,
+	})
+
+	// Subscribe if it is the first subscriber.
+	if len(h.subscriber[key]) == 1 {
+		h.Logger.Infof("subscribe because the first subscriber is joining")
+		ctx := context.Background()
+		err := pubsub.Subscribe(ctx, channelName)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return c, func() {
+		h.cleanupOne(cfg, credentials, channelName, id)
+	}, nil
+}
+
+func (h *Hub) getPubSub(cfg *config.RedisConfig, credentials *config.RedisCredentials) *redis.PubSub {
+	connKey := credentials.ConnKey()
+
+	pubsub, ok := h.pubsub[connKey]
+	if ok {
+		return pubsub
+	}
+
+	ctx := context.Background()
+	client := h.Pool.Client(cfg, credentials)
+	pubsub = client.Subscribe(ctx)
+	h.pubsub[connKey] = pubsub
+
+	go func() {
+		for {
+			iface, err := pubsub.ReceiveTimeout(ctx, RedisReadTimeout)
+			if os.IsTimeout(err) {
+				continue
+			}
+			if err != nil {
+				h.cleanupAll(cfg, credentials)
```

Comment:
This goroutine would exit on error when receiving. I suppose a context cancellation error is the normal exit condition; can we log other errors (i.e. abnormal exit)?